src/main/java/com/yelp/nrtsearch/server/luceneserver/search/query/multifunction/GeoPointDecayFilterFunction.java

```diff
@@ -0,0 +1,207 @@
+   * @param indexState indexState for validation and doc value lookup
+   */
+  public GeoPointDecayFilterFunction(
+      Query filterQuery,
+      float weight,
+      MultiFunctionScoreQuery.DecayFunction decayFunction,
+      IndexState indexState) {
+    super(filterQuery, weight, decayFunction);
+    this.decayFunction = decayFunction;
+    this.fieldName = decayFunction.getFieldName();
+    this.decayType = getDecayType(decayFunction.getDecayType());
+    this.origin = decayFunction.getGeoPoint();
+    this.decay = decayFunction.getDecay();
+    double userGivenScale = GeoUtils.getDistance(decayFunction.getScale());
+    this.scale = decayType.computeScale(userGivenScale, decay);
+    this.offset =
+        !decayFunction.getOffset().isEmpty()
+            ? GeoUtils.getDistance(decayFunction.getOffset())
+            : 0.0;
+    this.indexState = indexState;
+    validateLatLonField(indexState.getField(fieldName));
+  }
+
+  public void validateLatLonField(FieldDef fieldDef) {
+    if (!LAT_LON.equals(fieldDef.getType())) {
+      throw new IllegalArgumentException(
+          fieldName
+              + " should be a LAT_LON to apply geoPoint decay function but it is: "
+              + fieldDef.getType());
+    }
+    LatLonFieldDef latLonFieldDef = (LatLonFieldDef) fieldDef;
+    // TODO: Add support for multi-value fields
+    if (latLonFieldDef.isMultiValue()) {
+      throw new IllegalArgumentException(
+          "Multivalued fields are not supported for decay functions yet");
+    }
+    if (!latLonFieldDef.hasDocValues()) {
+      throw new IllegalStateException("No doc values present for LAT_LON field: " + fieldName);
+    }
+  }
+
+  @Override
+  public LeafFunction getLeafFunction(LeafReaderContext leafContext) {
+    return new GeoPointDecayLeafFunction(leafContext);
+  }
+
+  public final class GeoPointDecayLeafFunction implements LeafFunction {
+
+    SegmentDocLookup segmentDocLookup;
+
+    public GeoPointDecayLeafFunction(LeafReaderContext context) {
+      segmentDocLookup = indexState.docLookup.getSegmentLookup(context);
+    }
+
+    @Override
+    public double score(int docId, float innerQueryScore) throws IOException {
+      segmentDocLookup.setDocId(docId);
+      if (!validateDocValuesPresent()) {
+        return 0.0;
+      } else {
+        LoadedDocValues<GeoPoint> geoPointLoadedDocValues =
+            (LoadedDocValues<GeoPoint>) segmentDocLookup.get(fieldName);
+        GeoPoint latLng = geoPointLoadedDocValues.get(0);
+        double distance =
+            GeoUtils.arcDistance(
+                origin.getLatitude(), origin.getLongitude(), latLng.getLat(), latLng.getLon());
+        double score = decayType.computeScore(distance, offset, scale);
+        return score * getWeight();
+      }
+    }
+
+    public boolean validateDocValuesPresent() {
+      return segmentDocLookup.containsKey(fieldName) && !segmentDocLookup.get(fieldName).isEmpty();
+    }
```

Comment:
You shouldn't need to do `containsKey`, since you already checked that doc values are enabled. You will also want to avoid multiple calls to `segmentDocLookup.get()`, since it will reload data from the index. Something like this should work
```
LoadedDocValues<GeoPoint> geoPointLoadedDocValues =
            (LoadedDocValues<GeoPoint>) segmentDocLookup.get(fieldName);
if (geoPointLoadedDocValues.isEmpty()) {
  return 0.0;
} else {
  ...
}
```